src/ros_filter.cpp

```diff
@@ -391,71 +391,71 @@ namespace RobotLocalization
              "Integration time is " << std::setprecision(20) << currentTime << "\n"
              << measurementQueue_.size() << " measurements in queue.\n");
 
-    // If we have any measurements in the queue, process them
-    if(measurementsReady_.timed_wait(lock,boost::posix_time::milliseconds(filter_.getSensorTimeout()*1000)))
+    // If we do not have measurements in the queue wait until it is time to publish
+    if(measurementQueue_.empty())
     {
-      if(!measurementQueue_.empty())
+      if(!measurementsReady_.timed_wait(lock, boost::posix_time::milliseconds(1000*(timeToPublish_.left()))))
       {
 
-        double lastMessageTime;
-        while (!measurementQueue_.empty())
+        if (filter_.getInitializedStatus())
         {
-          Measurement measurement = measurementQueue_.top();
-          measurementQueue_.pop();
-          bool queueEmpty = measurementQueue_.empty();
-          double nextMeasurementTime = 0;
-          if(!queueEmpty)
+          // In the event that we don't get any measurements for a long time,
+          // we still need to continue to estimate our state. Therefore, we
+          // should project the state forward here.
+          double lastUpdateDelta = ros::Time::now().toSec() - filter_.getLastUpdateTime();
+
+          // If we get a large delta, then continuously predict until
+          if (lastUpdateDelta >= filter_.getSensorTimeout())
           {
-            nextMeasurementTime = measurementQueue_.top().time_;
+            RF_DEBUG("Sensor timeout! Last update time was " << filter_.getLastUpdateTime() <<
+                     ", current time is " << currentTime <<
+                     ", delta is " << lastUpdateDelta << "\n");
+
+            filter_.validateDelta(lastUpdateDelta);
+            filter_.predict(lastUpdateDelta);
+
+            // Update the last measurement time and last update time
+            filter_.setLastMeasurementTime(filter_.getLastMeasurementTime() + lastUpdateDelta);
+            filter_.setLastUpdateTime(filter_.getLastUpdateTime() + lastUpdateDelta);
```

Comment:
You call `getLastUpdateTime()` 3 times before here (including this one). Consider adding a `const TYPE lastUpdateTime` variable